File: ss-won/programmers/42898.cpp
#include <string>
#include <vector>
using namespace std;
const int MAX = 501;
const int dv = 1000000007;
int cache[MAX][MAX];
char map[MAX][MAX];

int getPath(int x, int y, int m, int n)
{
    if (x > m || y > n || map[x][y] == '1')
        return 0;
    if (cache[x][y] != -1)
        return cache[x][y];
    if (x == m && y == n)
        return cache[x][y] = 1;
    return cache[x][y] = (getPath(x + 1, y, m, n) + getPath(x, y + 1, m, n)) % dv;
}

int solution(int m, int n, vector<vector<int>> puddles)
{
    int answer = 0;
    // cache 초기화
    for (int i = 0; i <= m; i++)
    {
        fill(&cache[i][0], &cache[i][n] + 1, -1);
    }

    // puddles 표시
    for (auto puddle : puddles)
    {
        int x = puddle[0], y = puddle[1];
        map[x][y] = '1';
    }

    answer = getPath(1, 1, m, n);
    return answer;
}

File: ss-won/programmers/42898.py
def solution(m, n, puddles):
    answer = 0
    cache = [[-1 for _ in range(n+1)] for _ in range(m+1)]
    matrix = [[0 for _ in range(n+1)] for _ in range(m+1)]
    for puddle in puddles:
        x, y = puddle[0], puddle[1]
        matrix[x][y] = 1

    def getPath(x, y):
        if x > m or y > n or matrix[x][y] == 1:
            return 0
        if cache[x][y] != -1:
            return cache[x][y]
        if x == m and y == n:
            return 1
        cache[x][y] = (getPath(x+1, y) + getPath(x, y+1)) % 1000000007
        return cache[x][y]
    answer = getPath(1, 1)
    return answer
